File: netDxf/Entities/MLeaderContext.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using netDxf.Header;

namespace netDxf.Entities
{
    public sealed partial class MLeaderContext
    {
        private MLeaderMTextContent mtext;
        private MLeaderBlockContent block;
        private readonly MLeaderChildCollection<MLeaderNode> leaders;
        /// <summary>Creates an empty context with the published coordinate defaults.</summary>
        public MLeaderContext() { this.leaders = new MLeaderChildCollection<MLeaderNode>(this); }
        /// <summary>Gets the ordered leader branches. Their indices are stored values, not array offsets.</summary>
        public Collection<MLeaderNode> Leaders { get { return this.leaders; } }
        /// <summary>Gets or sets embedded text content. Remove block content before assigning text.</summary>
        public MLeaderMTextContent MText
        {
            get { return this.mtext; }
            set
            {
                if (ReferenceEquals(value,this.mtext)) return;
                if (value != null && (value.Parent != null || this.block != null)) throw new ArgumentException("Content must be unowned and mutually exclusive.",nameof(value));
                value?.CheckDocument(this.Document);
                if (this.mtext != null) this.mtext.Parent=null;
                this.mtext=value; if(value!=null)value.Parent=this;
            }
        }
        /// <summary>Gets or sets embedded block content. Remove text content before assigning a block.</summary>
        public MLeaderBlockContent Block
        {
            get { return this.block; }
            set
            {
                if (ReferenceEquals(value,this.block)) return;
                if (value != null && (value.Parent != null || this.mtext != null)) throw new ArgumentException("Content must be unowned and mutually exclusive.",nameof(value));
                value?.CheckDocument(this.Document);
                if (this.block != null) this.block.Parent=null;
                this.block=value; if(value!=null)value.Parent=this;
            }
        }
        internal override IEnumerable<MLeaderData> Children
        { get { if(this.mtext!=null)yield return this.mtext; if(this.block!=null)yield return this.block; foreach(var node in this.leaders)yield return node; } }
        internal override void CopyChildrenTo(MLeaderData target)
        {
            var copy=(MLeaderContext)target;
            if(this.mtext!=null)copy.MText=(MLeaderMTextContent)this.mtext.Clone();
            if(this.block!=null)copy.Block=(MLeaderBlockContent)this.block.Clone();
            foreach(var node in this.leaders)copy.Leaders.Add((MLeaderNode)node.Clone());
        }
        internal override void ValidateValues(DxfVersion version)
        {
            base.ValidateValues(version);
            ValidateDirection(this.PlaneXAxis); ValidateDirection(this.PlaneYAxis); ValidateDirection(Vector3.CrossProduct(this.PlaneXAxis,this.PlaneYAxis));
            this.mtext?.ValidateValues(version); this.block?.ValidateValues(version);
            foreach(var node in this.leaders)node.ValidateValues(version);
        }
    }
    public sealed partial class MLeaderMTextContent
    {
        private readonly Collection<double> columnHeights=new Collection<double>();
        /// <summary>Gets the ordered group-144 column heights, independent of the stored width and height fields.</summary>
        public Collection<double> ColumnHeights { get { return this.columnHeights; } }
        internal override void CopyChildrenTo(MLeaderData target) { foreach(double value in this.columnHeights)((MLeaderMTextContent)target).ColumnHeights.Add(value); }
        internal override void ValidateValues(DxfVersion version)
        {
            base.ValidateValues(version); ValidateDirection(this.Normal); ValidateDirection(this.Direction);
            if(this.Style==null)throw new InvalidOperationException("Embedded MULTILEADER text requires a registered STYLE reference.");
            if(this.ColumnType<0 || this.ColumnType>2)throw new NotSupportedException("Unsupported embedded MULTILEADER text column type.");
            foreach(double value in this.columnHeights) { Finite(value); if(value<0)throw new InvalidOperationException("Column heights cannot be negative."); }
        }
    }
    public sealed partial class MLeaderBlockContent
    {
        private readonly Collection<double> matrix=new Collection<double>();
        /// <summary>Gets the sixteen group-47 values in their original DXF order. Empty means no stored matrix.</summary>
        public Collection<double> TransformationMatrix { get { return this.matrix; } }
        internal override void CopyChildrenTo(MLeaderData target) { foreach(double value in this.matrix)((MLeaderBlockContent)target).TransformationMatrix.Add(value); }
        internal override void ValidateValues(DxfVersion version)
        {
            base.ValidateValues(version); ValidateDirection(this.Normal);
            if(this.Block==null)throw new InvalidOperationException("Embedded MULTILEADER block content requires a registered BLOCK_RECORD reference.");
            if(this.matrix.Count!=0 && this.matrix.Count!=16)throw new InvalidOperationException("A stored MULTILEADER block matrix must contain exactly sixteen values.");
            foreach(double value in this.matrix)Finite(value);
        }
    }
    /// <summary>A pair of stored WCS break endpoints.</summary>
    public sealed class MLeaderBreak
    {
        /// <summary>Creates a finite break pair without changing endpoint order.</summary>
        public MLeaderBreak(Vector3 start,Vector3 end) { MLeaderData.Finite(start);MLeaderData.Finite(end);this.Start=start;this.End=end; }
        /// <summary>Gets the start endpoint.</summary>
        public Vector3 Start {get;}
        /// <summary>Gets the end endpoint.</summary>
        public Vector3 End {get;}
    }
    /// <summary>An ordered group of line breaks associated with one stored segment index.</summary>
    public sealed class MLeaderLineBreaks
    {
        private readonly Collection<MLeaderBreak> breaks=new Collection<MLeaderBreak>();
        /// <summary>Gets or sets the stored group-90 index.</summary>
        public int Index {get;set;}
        /// <summary>Gets the ordered break pairs. Multiple pairs for the same index remain together.</summary>
        public Collection<MLeaderBreak> Breaks {get{return this.breaks;}}
        internal MLeaderLineBreaks Copy() {var copy=new MLeaderLineBreaks{Index=this.Index};foreach(var pair in this.breaks)copy.Breaks.Add(pair);return copy;}
    }
    public sealed partial class MLeaderNode
    {
        private readonly Collection<MLeaderBreak> breaks=new Collection<MLeaderBreak>();
        private readonly MLeaderChildCollection<MLeaderLine> lines;
        /// <summary>Creates a leader branch.</summary>
        public MLeaderNode() {this.lines=new MLeaderChildCollection<MLeaderLine>(this);}
        /// <summary>Gets ordered landing break pairs.</summary>
        public Collection<MLeaderBreak> Breaks {get{return this.breaks;}}
        /// <summary>Gets the ordered leader lines.</summary>
        public Collection<MLeaderLine> Lines {get{return this.lines;}}
        internal override IEnumerable<MLeaderData> Children {get{return this.lines;}}
        internal override void CopyChildrenTo(MLeaderData target) {var copy=(MLeaderNode)target;foreach(var pair in this.breaks)copy.Breaks.Add(pair);foreach(var line in this.lines)copy.Lines.Add((MLeaderLine)line.Clone());}
        internal override void ValidateValues(DxfVersion version) {base.ValidateValues(version);if(this.breaks.Any(b=>b==null))throw new InvalidOperationException("Break pairs cannot be null.");foreach(var line in this.lines)line.ValidateValues(version);}
    }
    public sealed partial class MLeaderLine
    {
        private readonly Collection<Vector3> vertices=new Collection<Vector3>();
        private readonly Collection<MLeaderLineBreaks> breaks=new Collection<MLeaderLineBreaks>();
        /// <summary>Gets stored WCS line vertices or spline fit points.</summary>
        public Collection<Vector3> Vertices {get{return this.vertices;}}
        /// <summary>Gets ordered indexed break groups.</summary>
        public Collection<MLeaderLineBreaks> Breaks {get{return this.breaks;}}
        internal override void CopyChildrenTo(MLeaderData target) {var copy=(MLeaderLine)target;foreach(var point in this.vertices)copy.Vertices.Add(point);foreach(var group in this.breaks)copy.Breaks.Add(group.Copy());}
        internal override void ValidateValues(DxfVersion version)
        {base.ValidateValues(version);foreach(var point in this.vertices)Finite(point);foreach(var group in this.breaks)if(group==null||group.Breaks.Count==0||group.Breaks.Any(b=>b==null))throw new InvalidOperationException("Line break groups require complete endpoint pairs.");}
    }
}

File: javascript/netDxf/Entities/MLeaderContext.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import {MLeaderData,MLeaderChildCollection,MLeaderComponentTypes} from './MLeaderData.js';
import {Vector3} from '../Vector3.js';
import {Copy} from '../../runtime/GeometryRuntime.js';
import {ReferenceList} from '../../runtime/ReferenceList.js';
import {ValueList} from '../../runtime/ValueList.js';
import {ArgumentException,InvalidOperationException,NotSupportedException,NullReferenceException} from '../../runtime/Errors.js';
// Empty Collection<T> uses the shared empty-array iterator; nonempty lists expose value defaults.
const emptyIterators=new Map();
class MLeaderValues extends ValueList{
  #zero;constructor(zero){super();this.#zero=zero;}
  IndexOf(item){return this.ToArray().findIndex(value=>value?.Equals?value.Equals(item):value===item||Number.isNaN(value)&&Number.isNaN(item));}
  GetEnumerator(){if(this.Count===0){const key=typeof this.#zero==='number'?'Double':'Vector3';if(!emptyIterators.has(key))emptyIterators.set(key,Object.freeze({get Current(){throw new InvalidOperationException('Enumeration has not started or has finished.');},MoveNext(){return false;},Reset(){},Dispose(){},next(){return{done:true,value:undefined};},[Symbol.iterator](){return this;}}));return emptyIterators.get(key);}const iterator=super.GetEnumerator(),owner=this;let active=false;return{
    get Current(){return active?iterator.Current:Copy(owner.#zero);},
    MoveNext(){active=iterator.MoveNext();return active;},Reset(){iterator.Reset();active=false;},Dispose(){iterator.Dispose();},
    next(){return this.MoveNext()?{done:false,value:this.Current}:{done:true,value:undefined};},[Symbol.iterator](){return this;}
  };}
}
export class MLeaderBreak{
  #start;#end;
  constructor(start,end){MLeaderData.Finite(start);MLeaderData.Finite(end);this.#start=Copy(start);this.#end=Copy(end);}
  get Start(){return Copy(this.#start);}get End(){return Copy(this.#end);}
}
export class MLeaderLineBreaks{
  Index=0;#breaks=new ReferenceList();get Breaks(){return this.#breaks;}
  Copy(){const copy=new MLeaderLineBreaks();copy.Index=this.Index;copy.Breaks.AddRange(this.#breaks);return copy;}
}
const state=new WeakMap();
export function InitializeMLeaderChildren(value,name){
  const s={};state.set(value,s);
  if(name==='MLeaderContext'){s.MText=null;s.Block=null;s.Leaders=new MLeaderChildCollection(value,'MLeaderNode');}
  if(name==='MLeaderMTextContent')s.ColumnHeights=new MLeaderValues(0);
  if(name==='MLeaderBlockContent')s.TransformationMatrix=new MLeaderValues(0);
  if(name==='MLeaderNode'){s.Breaks=new ReferenceList();s.Lines=new MLeaderChildCollection(value,'MLeaderLine');}
  if(name==='MLeaderLine'){s.Vertices=new MLeaderValues(Vector3.Zero);s.Breaks=new ReferenceList();}
  if(name==='MLeaderProperties'){s.ArrowHeads=new MLeaderChildCollection(value,'MLeaderArrowHead');s.BlockAttributes=new MLeaderChildCollection(value,'MLeaderBlockAttribute');}
}
export function InstallMLeaderChildren(Type,name){
  const collections={MLeaderContext:['Leaders'],MLeaderMTextContent:['ColumnHeights'],MLeaderBlockContent:['TransformationMatrix'],MLeaderNode:['Breaks','Lines'],MLeaderLine:['Vertices','Breaks'],MLeaderProperties:['ArrowHeads','BlockAttributes']}[name]??[];
  for(const key of collections)Object.defineProperty(Type.prototype,key,{get(){return state.get(this)[key];}});
  if(name==='MLeaderContext')for(const key of ['MText','Block'])Object.defineProperty(Type.prototype,key,{get(){return state.get(this)[key];},set(value){const s=state.get(this),old=s[key];if(old===value)return;if(value!==null){const T=MLeaderComponentTypes.get(key==='MText'?'MLeaderMTextContent':'MLeaderBlockContent');if(!T||!(value instanceof T)||value.Parent!==null||s[key==='MText'?'Block':'MText']!==null)throw new ArgumentException('Content must be unowned and mutually exclusive.','value');value.CheckDocument(this.Document);}if(old!==null)old.Parent=null;s[key]=value;if(value!==null)value.Parent=this;}});
  Object.defineProperty(Type.prototype,'Children',{get(){const s=state.get(this);if(name==='MLeaderContext')return{*[Symbol.iterator](){if(s.MText!==null)yield s.MText;if(s.Block!==null)yield s.Block;yield*s.Leaders;}};if(name==='MLeaderNode')return s.Lines;if(name==='MLeaderProperties')return{*[Symbol.iterator](){yield*s.ArrowHeads;yield*s.BlockAttributes;}};return [];}});
  Type.prototype.CopyChildrenTo=function(copy){const s=state.get(this);if(name==='MLeaderContext'){if(s.MText!==null)copy.MText=s.MText.Clone();if(s.Block!==null)copy.Block=s.Block.Clone();}
    for(const key of collections)for(const item of s[key]){if(key==='Breaks'&&name==='MLeaderLine'){if(item===null)throw new NullReferenceException();copy[key].Add(item.Copy());}else copy[key].Add(item instanceof MLeaderData?item.Clone():Copy(item));}
  };
  Type.prototype.ValidateValues=function(version){MLeaderData.prototype.ValidateValues.call(this,version);
    if(name==='MLeaderContext'){MLeaderData.ValidateDirection(this.PlaneXAxis);MLeaderData.ValidateDirection(this.PlaneYAxis);MLeaderData.ValidateDirection(Vector3.CrossProduct(this.PlaneXAxis,this.PlaneYAxis));this.MText?.ValidateValues(version);this.Block?.ValidateValues(version);for(const node of this.Leaders)node.ValidateValues(version);}
    if(name==='MLeaderMTextContent'){MLeaderData.ValidateDirection(this.Normal);MLeaderData.ValidateDirection(this.Direction);if(this.Style===null)throw new InvalidOperationException('Embedded MULTILEADER text requires a registered STYLE reference.');if(this.ColumnType<0||this.ColumnType>2)throw new NotSupportedException('Unsupported embedded MULTILEADER text column type.');for(const height of this.ColumnHeights){MLeaderData.Finite(height);if(height<0)throw new InvalidOperationException('Column heights cannot be negative.');}}
    if(name==='MLeaderBlockContent'){MLeaderData.ValidateDirection(this.Normal);if(this.Block===null)throw new InvalidOperationException('Embedded MULTILEADER block content requires a registered BLOCK_RECORD reference.');if(this.TransformationMatrix.Count!==0&&this.TransformationMatrix.Count!==16)throw new InvalidOperationException('A stored MULTILEADER block matrix must contain exactly sixteen values.');for(const cell of this.TransformationMatrix)MLeaderData.Finite(cell);}
    if(name==='MLeaderNode'){if([...this.Breaks].some(x=>x===null))throw new InvalidOperationException('Break pairs cannot be null.');for(const line of this.Lines)line.ValidateValues(version);}
    if(name==='MLeaderLine'){for(const point of this.Vertices)MLeaderData.Finite(point);for(const group of this.Breaks)if(group===null||group.Breaks.Count===0||[...group.Breaks].some(x=>x===null))throw new InvalidOperationException('Line break groups require complete endpoint pairs.');}
  };
}
